rpc/client.go

```diff
@@ -25,72 +24,91 @@ type UploadReportOptions struct {
 
 func UploadReport(ctx context.Context, opts UploadReportOptions) error {
 	if err := opts.convertAndUploadArtifacts(ctx); err != nil {
-		return errors.Wrap(err, "problem uploading tests for report")
+		return errors.Wrap(err, "uploading tests for report")
 	}
 	return errors.Wrap(uploadTests(ctx, gopb.NewCedarPerformanceMetricsClient(opts.ClientConn), opts.Report, opts.Report.Tests, opts.DryRun),
-		"problem uploading tests for report")
+		"uploading tests for report")
 }
 
 func (opts *UploadReportOptions) convertAndUploadArtifacts(ctx context.Context) error {
-	jobQueue := queue.NewLocalLimitedSize(runtime.NumCPU(), len(opts.Report.Tests)*2)
+	var wg sync.WaitGroup
+	catcher := grip.NewBasicCatcher()
+	testChan := make(chan poplar.Test)
+	wg.Add(1)
+	go opts.artifactProducer(ctx, testChan, catcher, &wg)
+
+	workers := 1
 	if !opts.SerializeUpload {
-		if err := jobQueue.Start(ctx); err != nil {
-			return errors.Wrap(err, "problem starting artifact upload queue")
-		}
-		defer jobQueue.Runner().Close(ctx)
+		workers = runtime.NumCPU()
 	}
-
-	queue := make([]poplar.Test, len(opts.Report.Tests))
-	for i, test := range opts.Report.Tests {
-		queue[i] = test
+	for i := 0; i < workers; i++ {
+		wg.Add(1)
+		go opts.artifactConsumer(ctx, testChan, catcher, &wg)
 	}
 
-	for len(queue) != 0 {
-		test := queue[0]
-		queue = queue[1:]
+	wg.Wait()
+	return catcher.Resolve()
+}
+
+func (opts *UploadReportOptions) artifactProducer(ctx context.Context, testChan chan poplar.Test, catcher grip.Catcher, wg *sync.WaitGroup) {
+	defer func() {
+		catcher.Add(recovery.HandlePanicWithError(recover(), nil, "artifact upload producer"))
+		close(testChan)
+		wg.Done()
+	}()
 
-		for _, subtest := range test.SubTests {
-			queue = append(queue, subtest)
+	testQueue := make([]poplar.Test, len(opts.Report.Tests))
+	for i, test := range opts.Report.Tests {
+		testQueue[i] = test
+	}
+	for len(testQueue) != 0 {
+		test := testQueue[0]
+		testQueue = testQueue[1:]
+		select {
+		case testChan <- test:
+		case <-ctx.Done():
+			return
 		}
 
-		for i := range test.Artifacts {
-			var err error
+		for _, subTest := range test.SubTests {
```

Comment:
Why can these not be appended in the for loop on L61? Is it because subtests must be uploaded after the test itself?

Reply:
Its because subtests can also have subtests, so we have to do a depth-first search. The uploading order actually doesn't really matter.